finagle-thrift/src/main/scala/com/twitter/finagle/thrift/MultiplexedService.scala

```diff
@@ -0,0 +1,117 @@
+
+class MultiplexedFinagleService(
+  services: Map[String, Service[Array[Byte], Array[Byte]]],
+  protocolFactory: TProtocolFactory,
+  maxThriftBufferSize: Int = Thrift.Server.maxThriftBufferSize)
+    extends Service[Array[Byte], Array[Byte]] {
+
+  private val serviceMap = services.mapValues(getFunctionMap)
+
+  def getFunctionMap(
+    service: Service[Array[Byte], Array[Byte]]
+  ): collection.Map[String, (TProtocol, Int) => Future[Array[Byte]]] = {
+    Try(scroogeFinagleServiceFunctionMap(service))
+      .orElse(Try(thriftFinagleServiceFunctionMap(service)))
+      .getOrElse(throw new IllegalArgumentException("%s cannot be multiplexed".format(service.getClass.getName)))
+  }
+
+  def scroogeFinagleServiceFunctionMap[T](target: AnyRef) = {
+    val m = target.getClass.getMethod("functionMap")
+    val accessible = m.isAccessible
+    m.setAccessible(true)
+    val result = m.invoke(target)
+      .asInstanceOf[collection.Map[String, (TProtocol, Int) => Future[Array[Byte]]]]
+    m.setAccessible(accessible)
+    result
+  }
+
+  def thriftFinagleServiceFunctionMap(target: AnyRef) = {
+    val f = target.getClass.getDeclaredField("functionMap")
+    val accessible = f.isAccessible
+    f.setAccessible(true)
+    val functionMap = f.get(target)
+      .asInstanceOf[java.util.Map[String, com.twitter.util.Function2[TProtocol, Integer, Future[Array[Byte]]]]]
+      .asScala
+      .mapValues(function => (prot: TProtocol, seqId: Int) => function(prot, seqId))
+    f.setAccessible(accessible)
+    functionMap
+  }
+
+  final def apply(request: Array[Byte]): Future[Array[Byte]] = {
+    val inputTransport = new TMemoryInputTransport(request)
+    val iprot = protocolFactory.getProtocol(inputTransport)
+    try {
+      val msg = iprot.readMessageBegin()
+      val index = msg.name.indexOf(TMultiplexedProtocol.SEPARATOR)
+      if (index == -1) {
+        exception(msg.name, msg.seqid, TApplicationException.PROTOCOL_ERROR,
+          s"This is a multiplexed service, with available service names: [${serviceMap.keys.mkString(", ")}]")
+      } else {
+        val serviceName = msg.name.substring(0, index)
+        val functionName = msg.name.substring(index + 1)
+        val functionMap = serviceMap.get(serviceName)
+        val func = functionMap.flatMap(_.get(functionName))
+        func match {
+          case Some(fn) =>
+            fn(iprot, msg.seqid)
+          case _ =>
+            TProtocolUtil.skip(iprot, TType.STRUCT)
+            val errorMsg = if (functionMap.isEmpty) {
+              s"""Invalid service name: ${serviceName}, available serivce names: [${serviceMap.keys.mkString(", ")}]"""
+            } else {
+              s"""Invalid method name: ${functionName}, available method names: [${functionMap.get.keys.mkString(", ")}]"""
+            }
+            exception(msg.name, msg.seqid, TApplicationException.UNKNOWN_METHOD, errorMsg)
+        }
+      }
+    } catch {
+      case e: Throwable => Future.exception(e)
```

Comment:
should probably only catch NonFatal's:

`case NonFatal(e) => Future.exception(e)`